Index: machdep.c
===================================================================
--- machdep.c   (revision 256766)
+++ machdep.c   (working copy)
@@ -1489,6 +1489,9 @@
 
        init_proc0(kernelstack.pv_va);
 
+       cpufunc_control(CPU_CONTROL_DC_ENABLE|CPU_CONTROL_IC_ENABLE,
+           CPU_CONTROL_DC_ENABLE|CPU_CONTROL_IC_ENABLE);
+
        arm_intrnames_init();
        arm_vector_init(ARM_VECTORS_HIGH, ARM_VEC_ALL);
        arm_dump_avail_init(memsize, sizeof(dump_avail) / sizeof(dump_avail[0]));

